shopfloor_single_product_transfer/services/single_product_transfer.py

```diff
@@ -550,31 +627,66 @@ def _find_user_move_line(self):
 
-    def _set_quantity__by_product(self, move_line, barcode, confirmation=False):
-        product_handlers = [
-            self._set_quantity__scan_product,
-            self._set_quantity__scan_packaging,
-            self._set_quantity__scan_lot,
-        ]
-        product_response = self._use_handlers(product_handlers, move_line, barcode)
-        if product_response:
-            return product_response
-
-    def _set_quantity__by_location(self, move_line, barcode, confirmation=False):
-        search = self._actions_for("search")
-        location = search.location_from_scan(barcode)
-        handlers = [
-            # Cannot confirm if qty_done is not valid (> qty todo)
-            self._set_quantity__check_quantity_done,
+    def _set_quantity__by_location_handlers(self):
+        return [
             self._set_quantity__check_location,
             self._set_quantity__post_move,
         ]
-        if location:
+
+    def _set_quantity__by_location(self, move_line, location, confirmation=False):
+        # We're about to leave the `set_quantity` screen.
+        # First ensure that quantity is valid.
+        invalid_qty_response = self._set_quantity__check_quantity_done(move_line)
+        if invalid_qty_response:
+            return invalid_qty_response
+        move_line.result_package_id = False
+        handlers = self._set_quantity__by_location_handlers()
+        response = self._use_handlers(
+            handlers, move_line, location=location, confirmation=confirmation
+        )
+        if response:
+            return response
+
+    def _set_quantity__by_package(self, move_line, package, confirmation=False):
+        # We're about to leave the `set_quantity` screen.
+        # First ensure that quantity is valid.
+        invalid_qty_response = self._set_quantity__check_quantity_done(move_line)
+        if invalid_qty_response:
+            return invalid_qty_response
+        # If package isn't empty, then check its location then post the move
+        if package.quant_ids:
+            location = package.location_id
+            handlers = self._set_quantity__by_location_handlers()
             response = self._use_handlers(
-                handlers, location, move_line, confirmation=confirmation
+                handlers, move_line, location, confirmation=confirmation
             )
-            if response:
-                return response
+            move_line.result_package_id = package
+            return response
+        # Else, go to `set_location` screen
+        move_line.result_package_id = package
+        return self._response_for_set_location(move_line, package)
+
+    def _scan_package(self, package):
+        handlers = [
+            self._scan_package__check_location,
+        ]
+        response = self._use_handlers(handlers, package)
+        if response:
+            return response
+        return self._response_for_select_product(package=package)
```

Comment:
I would display the location & package on the screen
```suggestion
        return self._response_for_select_product(package=package, location=package.location_id)
```